File: mars4/src/main/java/edu/missouristate/mars/mips/dump/SegmentWindowDumpFormat.java
package edu.missouristate.mars.mips.dump;

import edu.missouristate.mars.Globals;
import edu.missouristate.mars.ProgramStatement;
import edu.missouristate.mars.Settings;
import edu.missouristate.mars.mips.hardware.AddressErrorException;
import edu.missouristate.mars.mips.hardware.Memory;
import edu.missouristate.mars.util.Binary;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.PrintStream;

/**
 * Dump MIPS memory contents in Segment Window format.  Each line of
 * text output resembles the Text Segment Window or Data Segment Window
 * depending on which segment is selected for the dump.  Written
 * using PrintStream's println() method.  Each line of Text Segment
 * Window represents one word of text segment memory.  The line
 * includes (1) address, (2) machine code in hex, (3) basic instruction,
 * (4) source line.  Each line of Data Segment Window represents 8
 * words of data segment memory.  The line includes address of first
 * word for that line followed by 8 32-bit values.
 * <p>
 * In either case, addresses and values are displayed in decimal or
 * hexadecimal representation according to the corresponding settings.
 *
 * @author Pete Sanderson
 * @version January 2008
 */
public class SegmentWindowDumpFormat extends AbstractDumpFormat {
    /**
     * Constructor.  There is no standard file extension for this format.
     */
    public SegmentWindowDumpFormat() {
        super("Text/Data Segment Window", null, " Text Segment Window or Data Segment Window format to text file", "txt");
    }

    /**
     * Write MIPS memory contents in Segment Window format.  Each line of
     * text output resembles the Text Segment Window or Data Segment Window
     * depending on which segment is selected for the dump.  Written
     * using PrintStream's println() method.
     *
     * @param file         File in which to store MIPS memory contents.
     * @param firstAddress first (lowest) memory address to dump.  In bytes but
     *                     must be on word boundary.
     * @param lastAddress  last (highest) memory address to dump.  In bytes but
     *                     must be on word boundary.  Will dump the word that starts at this address.
     * @throws AddressErrorException if firstAddress is invalid or not on a word boundary.
     * @throws IOException           if error occurs during file output.
     */
    public void dumpMemoryRange(@NotNull File file, int firstAddress, int lastAddress)
            throws AddressErrorException, IOException {

        PrintStream out = new PrintStream(new FileOutputStream(file));

        boolean hexAddresses = Globals.getSettings().getBooleanSetting(Settings.DISPLAY_ADDRESSES_IN_HEX);

        // If address in data segment, print in same format as Data Segment Window
        if (Memory.inDataSegment(firstAddress)) {
            boolean hexValues = Globals.getSettings().getBooleanSetting(Settings.DISPLAY_VALUES_IN_HEX);
            int offset = 0;
            StringBuilder string = new StringBuilder();
            try {
                for (int address = firstAddress; address <= lastAddress; address += Memory.WORD_LENGTH_BYTES) {
                    if (offset % 8 == 0) {
                        string = new StringBuilder(((hexAddresses) ? Binary.intToHexString(address) : Binary.unsignedIntToIntString(address)) + "    ");
                    }
                    offset++;
                    Integer temp = Globals.memory.getRawWordOrNull(address);
                    if (temp == null)
                        break;
                    string.append((hexValues)
                            ? Binary.intToHexString(temp)
                            : ("           " + temp).substring(temp.toString().length())).append(" ");
                    if (offset % 8 == 0) {
                        out.println(string);
                        string = new StringBuilder();
                    }
                }
            } finally {
                out.close();
            }
            return;
        }

        if (!Memory.inTextSegment(firstAddress)) {
            return;
        }
        // If address in text segment, print in same format as Text Segment Window
        out.println(" Address    Code        Basic                     Source");
        //           12345678901234567890123456789012345678901234567890
        //                    1         2         3         4         5
        out.println();
        String string;
        try {
            for (int address = firstAddress; address <= lastAddress; address += Memory.WORD_LENGTH_BYTES) {
                string = ((hexAddresses) ? Binary.intToHexString(address) : Binary.unsignedIntToIntString(address)) + "  ";
                Integer temp = Globals.memory.getRawWordOrNull(address);
                if (temp == null)
                    break;
                string += Binary.intToHexString(temp) + "  ";
                try {
                    ProgramStatement ps = Globals.memory.getStatement(address);
                    string += (ps.getPrintableBasicAssemblyStatement() + "                      ").substring(0, 22);
                    string += (((ps.getSource().equals("")) ? "" : Integer.valueOf(ps.getSourceLine()).toString()) + "     ").substring(0, 5);
                    string += ps.getSource();
                } catch (AddressErrorException ignored) {
                }
                out.println(string);
            }
        } finally {
            out.close();
        }
    }
}

File: mars5/src/main/kotlin/edu/missouristate/mars/mips/dump/SegmentWindowDumpFormat.kt
/*
 * Copyright (c) 2003-2024, Pete Sanderson and Kenneth Vollmar
 * Copyright (c) 2024-present, Nicholas Hubbard
 *
 * Created by Pete Sanderson (psanderson@otterbein.edu) and Kenneth Vollmar (kenvollmar@missouristate.edu)
 * Maintained by Nicholas Hubbard (nhubbard@users.noreply.github.com)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * 1. The above copyright notice and this permission notice shall be included in all copies or substantial portions of
 *    the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 *
 * Copyright (c) 2017-2024, Niklas Persson
 * Copyright (c) 2024-present, Nicholas Hubbard
 *
 * The IntelliJ plugin is licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for specific
 * language governing permissions and limitations under the License.
 */

package edu.missouristate.mars.mips.dump

import edu.missouristate.mars.CoreSpec
import edu.missouristate.mars.Globals
import edu.missouristate.mars.mips.hardware.AddressErrorException
import edu.missouristate.mars.mips.hardware.Memory
import edu.missouristate.mars.util.Binary
import java.io.File
import java.io.FileOutputStream
import java.io.PrintStream

class SegmentWindowDumpFormat : AbstractDumpFormat(
    "Text/Data Segment Window",
    "SegmentWindow",
    "Text Segment or Data Segment Window format as text file",
    "txt"
) {
    override fun dumpMemoryRange(file: File?, firstAddress: Int, lastAddress: Int) {
        val hexAddresses = Globals.config[CoreSpec.displayAddressesInHex]
        val hexValues = Globals.config[CoreSpec.displayValuesInHex]
        file?.let { f ->
            FileOutputStream(f).use { out ->
                PrintStream(out).use {
                    if (Memory.inDataSegment(firstAddress)) {
                        var offset = 0
                        var string = StringBuilder()
                        for (address in firstAddress..lastAddress step Memory.WORD_LENGTH_BYTES) {
                            if (offset % 8 == 0)
                                string = StringBuilder("${if (hexAddresses) Binary.intToHexString(address) else Binary.unsignedIntToIntString(address)}    ")
                            offset++
                            val temp = Globals.memory.getRawWordOrNull(address) ?: break
                            string.append(if (hexValues) Binary.intToHexString(temp) else "           $temp".substring(temp.toString().length)).append(" ")
                            if (offset % 8 == 0) {
                                it.println(string)
                                string = StringBuilder()
                            }
                        }
                        return
                    }
                    if (!Memory.inTextSegment(firstAddress)) return
                    it.println(" Address    Code        Basic                     Source")
                    it.println()
                    var string: String
                    for (address in firstAddress..lastAddress step Memory.WORD_LENGTH_BYTES) {
                        string = "${if (hexAddresses) Binary.intToHexString(address) else Binary.unsignedIntToIntString(address)}  "
                        val temp = Globals.memory.getRawWordOrNull(address) ?: break
                        string += "${Binary.intToHexString(temp)}  "
                        try {
                            val ps = Globals.memory.getStatement(address)
                            string += "${ps?.getPrintableBasicAssemblyStatement()}                      ".substring(0, 22)
                            string += ((if (ps?.getSource() == "") "" else Integer.valueOf(ps?.getSourceLine() ?: 0).toString()) + "     ").substring(0, 5)
                            string += ps?.getSource()
                        } catch (ignored: AddressErrorException) {}
                        it.println(string)
                    }
                }
            }
        }
    }
}